<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good)</span></div>

## 1 Interrupting the flow

In [5]:
number=0

while True:
    print(number)
    number += 1
    if number > 4: break
        
#condition is <5

#OR

number = 0

while number < 5:
    print(number)
    number += 1

0
1
2
3
4
0
1
2
3
4


## 2 List comprehension!

### 2.1 Basic syntax

In [9]:
[number for number in range(10)]

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [10]:
[number**2 for number in range(10)]

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

### 2.2 List comprehension with conditions

In [11]:
[number for number in range(10) if number % 2 ==0]
#even numbers

[0, 2, 4, 6, 8]

## 3 Other useful stuff

### 3.1 for with unpacking

In [12]:
x, y, z=[1, 2, 3]
print(f'x = {x}, y = {y}, z = {z}')

x = 1, y = 2, z = 3


In [16]:
py_superhero_info=[['Natasha Romanoff', 'Black Widow'],
                     ['Tony Stark', 'Iron Man'],
                     ['Stephen Strange', 'Doctor Strange']]

for real_name, super_name in py_superhero_info:
    print(f"{real_name} is Marvel's {super_name}!")
    
#real_name is key
#super_name is value

Natasha Romanoff is Marvel's Black Widow!
Tony Stark is Marvel's Iron Man!
Stephen Strange is Marvel's Doctor Strange!


### 3.2 for with a zip()

In [14]:
super_names = ["Black Widow", "Iron Man", "Doctor Strange"]
real_names = ["Natasha Romanoff", "Tony Stark", "Stephen Strange"]

for real_name, super_name in zip(real_names,super_names):
    print(f"{real_name} is Marvel's {super_name}!")
    
#2 separate lists

Natasha Romanoff is Marvel's Black Widow!
Tony Stark is Marvel's Iron Man!
Stephen Strange is Marvel's Doctor Strange!


### 3.3 for with dictionaries

In [15]:
superhero_info={"Natasha Romanoff": "Black Widow",
                "Tony Stark": "Iron Man",
                "Stephen Strange": "Doctor Strange"}

for key, value in superhero_info.items():
    print(f"{key} is Marvel's {value}!")

Natasha Romanoff is Marvel's Black Widow!
Tony Stark is Marvel's Iron Man!
Stephen Strange is Marvel's Doctor Strange!


## Exercise 1 :  Make me an odd list

In [22]:
%%timeit
num_list = []
for i in range(10):
    if i % 2 == 0:
        continue
    num_list.append(i**2)

print(num_list)


[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 4

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 4

## Exercise 2 :  Make me another odd list

In [24]:
%%timeit
[number**2 for number in range(1,10,2)]

1.11 µs ± 5.32 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


## Exercise 3 :  Time me!

in exercise 1:
2.87 µs ± 26.4 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

in exercise 2:
1.11 µs ± 5.32 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

exercise 2 is faster
